client/app/components/tags-control/TagsControl.jsx

```diff
@@ -0,0 +1,103 @@
+import { map, trim, isObject, isFunction } from 'lodash';
+import React from 'react';
+import PropTypes from 'prop-types';
+
+// eslint-disable-next-line import/prefer-default-export
+export class TagsControl extends React.Component {
+  static propTypes = {
+    tags: PropTypes.arrayOf(PropTypes.string),
+    canEdit: PropTypes.bool,
+    getAvailableTags: PropTypes.func,
+    onEdit: PropTypes.func,
+  };
+
+  static defaultProps = {
+    tags: [],
+    canEdit: false,
+    getAvailableTags: () => {},
+    onEdit: () => {},
+  };
+
+  openEditTagsModal(tags, availableTags = []) {
+    const { $uibModal } = this.props; // eslint-disable-line react/prop-types
+
+    $uibModal
+      .open({
+        component: 'tagsEditorModal',
+        resolve: {
+          tags: () => tags,
+          availableTags: () => availableTags,
+        },
+      }).result.then((newTags) => {
+        const { onEdit } = this.props;
+        if (isFunction(onEdit)) {
+          onEdit(newTags);
+        }
+      });
+  }
+
+  editTags() {
+    const { getAvailableTags } = this.props;
+    const tags = map(this.props.tags, trim);
+
+    if (isFunction(getAvailableTags)) {
```

Comment:
Why wouldn't it be a function? Same goes for all the protective code around the result from calling this function, instead of assuming it returns a Promise.